File: python/redact_secret_adapters/mask_leaf.py
"""The single primitive every adapter builds on: mask one leaf string with
an injected ``scan_and_redact``, and never let a core failure or a
``block`` finding put text on the wire.

Mirrors ``packages/adapter/src/mask-leaf.ts`` line for line so the two
languages make the same decision given the same finding.
"""

from __future__ import annotations

from typing import Any, Callable, Optional

BLOCK_MARKER = "[REDACTED:BLOCKED]"
ERROR_MARKER = "[REDACTED:ERROR]"
LIMIT_MARKER = "[REDACTED:LIMIT_EXCEEDED]"
CYCLE_MARKER = "[REDACTED:CYCLE]"

# Bounds enforced by every walker, the logging filter, and the span
# processor. A field that exceeds max_string_length, or a value reached
# only after max_depth/max_array_length/max_object_keys/max_total_leaves is
# spent, never reaches the core: it becomes LIMIT_MARKER instead. Every log
# call pays the scan cost, so these bounds also cap per-call latency for a
# pathological extra mapping.
DEFAULT_LIMITS: dict[str, int] = {
    "max_depth": 8,
    "max_array_length": 1000,
    "max_object_keys": 200,
    "max_string_length": 200_000,
    "max_total_leaves": 5000,
}


def mask_leaf_with(
    scan_and_redact: Callable[..., Any],
    text: str,
    *,
    policy: Optional[Any] = None,
    max_string_length: Optional[int] = None,
) -> str:
    """Masks one leaf string.

    Any exception the core raises (``redact_secret.SecretScanError`` and
    its subclasses, or anything else a misbehaving ``policy`` raises)
    fails closed: the leaf becomes ``ERROR_MARKER``, never the original
    text and never the exception's own message. A ``block`` finding
    replaces the entire leaf with ``BLOCK_MARKER``: ``scan_and_redact``
    already substitutes ``block`` findings in place like ``redact`` ones,
    but an inline placeholder still leaves the rest of the string visible,
    which is not the documented host decision for a block-worthy secret.
    For a plain string log call the leaf *is* the message, and for a
    formatted field or an ``exc_text`` the leaf is that field's whole text.
    """
    if not isinstance(text, str):
        raise TypeError("mask_leaf_with: text must be a str")

    limit = max_string_length if max_string_length is not None else DEFAULT_LIMITS["max_string_length"]
    if len(text) > limit:
        return LIMIT_MARKER

    try:
        result = scan_and_redact(text, policy)
    except Exception:
        return ERROR_MARKER

    if any(finding.action == "block" for finding in result.findings):
        return BLOCK_MARKER
    return result.text

File: packages/adapter/src/mask-leaf.ts
/**
 * The single primitive every adapter builds on: mask one leaf string with
 * an injected `scanAndRedact`, and never let a core failure or a `block`
 * finding put text on the wire.
 */

import type { SecretAction } from "@redact-secret/core";

import type { Limits, MaskLeafOptions, ScanAndRedact } from "./types.js";

export const BLOCK_MARKER = "[REDACTED:BLOCKED]";
export const ERROR_MARKER = "[REDACTED:ERROR]";
export const LIMIT_MARKER = "[REDACTED:LIMIT_EXCEEDED]";
export const CYCLE_MARKER = "[REDACTED:CYCLE]";

const BLOCK: SecretAction = "block";

/**
 * Bounds enforced by every walker and the span processor. A field that
 * exceeds `maxStringLength`, or a value reached only after `maxDepth`/
 * `maxArrayLength`/`maxObjectKeys`/`maxTotalLeaves` is spent, never
 * reaches the core: it becomes {@link LIMIT_MARKER} instead. Every log
 * call pays the scan cost, so these bounds also cap per-call latency for
 * pathological merging objects.
 */
export const DEFAULT_LIMITS: Limits = Object.freeze({
  maxDepth: 8,
  maxArrayLength: 1000,
  maxObjectKeys: 200,
  maxStringLength: 200_000,
  maxTotalLeaves: 5000,
});

/**
 * Masks one leaf string. Any thrown error — including `NOT_INITIALIZED` if
 * a host skipped `await initialize()` — fails closed: the leaf becomes
 * {@link ERROR_MARKER}, never the original text and never the error's own
 * message. A `block` finding replaces the entire leaf with
 * {@link BLOCK_MARKER}: `scanAndRedact` already substitutes `block`
 * findings in place like `redact` ones, but an inline placeholder still
 * leaves the rest of the string visible, which is not the documented host
 * decision for a block-worthy secret. For a plain string log call the leaf
 * *is* the message, and for a merged field or an `err.message` the leaf is
 * that field's whole text.
 */
export function maskLeafWith(
  scanAndRedact: ScanAndRedact,
  text: string,
  { policy, maxStringLength }: MaskLeafOptions = {},
): string {
  if (typeof text !== "string") {
    throw new TypeError("maskLeafWith: text must be a string");
  }
  const limit = maxStringLength ?? DEFAULT_LIMITS.maxStringLength;
  if (text.length > limit) return LIMIT_MARKER;

  let result: ReturnType<ScanAndRedact>;
  try {
    result = scanAndRedact(text, { policy });
  } catch {
    return ERROR_MARKER;
  }
  if (result.findings.some((finding) => finding.action === BLOCK)) {
    return BLOCK_MARKER;
  }
  return result.text;
}
